### ISSUE TITLE
Computation of 2nd order gradient issue when converting nodal field to edge attribute field

### ISSUE THREAD

**Issue by jahawk** [rusty1s/pytorch_geometric#1807]
## ❓ Questions & Help
Hi,

I've being trying to implement a Graph model using the ARMAConv layer in which I use nodal distances to compute the edge_attributes as weights and then compute the gradients of resulting fields w.r.t the coordinates after passing through the GN layer but I'm having problems with the 2nd order derivative. 

Don't now if this is something you might have already seen so wanted to ask for some guidance,

Simpler with a minimal code without any layer to check the issue:
Note: At the end my GN gives me back also a Nodal field from which I expected to compute the gradients. Since it wasn't working for the 2nd order derivative I tried the following simple code

##
import torch
from torch import Tensor
# Define coordinates, field and edges connectivity
x = torch.tensor( [[0.0,0.0,0.0],
                  [1.0,0.0,0.0],
                  [0.0,1.0,0.0],
                  [0.0,0.0,1.0],
                  [1.0,1.0,1.0]], requires_grad=True)
field = torch.tensor([[ 0.],
                      [ 1.],
                      [ 4.],
                      [ 9.],
                      [25.]], requires_grad=True)
edges = torch.tensor(
    [[0, 0, 0, 1, 1, 1, 1, 2, 2, 2, 2, 3, 3, 3, 3, 4, 4, 4],
     [1, 2, 3, 0, 2, 3, 4, 0, 1, 3, 4, 0, 1, 2, 4, 1, 2, 3]])

# Compute weights from nodal distances
row, col = edges
dist_sq = torch.sum( (x[col] - x[row]).pow(2) , dim = 1)
dist_sq.retain_grad()
weight = dist_sq.pow(-1)

# Compute first gradient w.r.t x // this will show some values
ones = torch.ones_like(weight)
qx = torch.autograd.grad(weight, x, grad_outputs=ones, create_graph=True, retain_graph=True)[0]
qx.retain_grad()
print(qx)

# Compute second gradient w.r.t x // this will show zero and also prompt a UserWarning sayin that the .grad attribute of a non leaf tensor is being accesed so that I should use .retain_grad() on the non leaf tensor. Which I used on dist_sq and qx but doesn't seem to change the behaviour
ones = torch.ones_like(qx)
qxx = torch.autograd.grad(qx, x, grad_outputs=ones, create_graph=True, retain_graph=True)[0]
print(qxx)
##

If instead of creating an edge_attribute field I create a nodal field from x, say weight = x.pow(2) I will obtain the proper gradients. Any ideas?

Thanks in advance for any help

Extra info:
OS: Ubuntu 18
Nvidia docker 20.08
Python 3.6.1
Pytorch 1.7.0
Pytorch Geometric 1.6.1

**Comment by rusty1s** [comment]
This might be better asked in the PyTorch forums, since I'm not familiar with computing 2nd order gradients in PyTorch. However, I think your code is absolutely correct. I did not need to use `retain_grad` (there are no warnings for me):
```python
row, col = edges
dist_sq = torch.sum((x[col] - x[row]).pow(2), dim=1)
weight = dist_sq.pow(-1)

ones = torch.ones_like(weight)
qx = torch.autograd.grad(weight, x, grad_outputs=ones, create_graph=True)[0]
print(qx)

ones = torch.ones_like(qx)
qxx = torch.autograd.grad(qx, x, grad_outputs=ones, create_graph=True)[0]
print(qxx)
```
However, `qxx` is indeed zero, not exactly sure why :(

**Comment by jahawk** [comment]
Hi @rusty1s  thanks for the help! I will try at the pytorch forum and if I manage to understand and solve the issue I'll link it here later

**Comment by jahawk** [comment]
Hi again, I managed to understand the issue and solve for the computation of the second order gradients, here the [link](https://discuss.pytorch.org/t/second-order-gradient-zeroing-on-different-shape-tensor/102141/4) to the discussion at the pytorch forum
